File: src/domain/repositories/issue-repository.ts
import type { Issue } from '../entities/issue.js';
import type { IssueId } from '../value-objects/issue-id.js';
import type { ProjectId } from '../value-objects/project-id.js';

export interface IssueRepository {
  findById: (_id: IssueId) => Promise<Issue | null>;
  findByProjectId: (_projectId: ProjectId) => Promise<Issue[]>;
  save: (_issue: Issue) => Promise<void>;
  saveToProject: (_issue: Issue, _projectId: ProjectId) => Promise<void>;
}

File: src/main/kotlin/io/spiralhouse/jcvd/domain/repositories/IssueRepository.kt
package com.spiralhouse.jcvd.domain.repositories

import com.spiralhouse.jcvd.domain.entities.Issue
import com.spiralhouse.jcvd.domain.valueobjects.IssueId
import com.spiralhouse.jcvd.domain.valueobjects.IssueStatus
import com.spiralhouse.jcvd.domain.valueobjects.IssueType
import com.spiralhouse.jcvd.domain.valueobjects.ProjectId

interface IssueRepository {
    suspend fun findById(id: IssueId): Issue?
    suspend fun findByProject(projectId: ProjectId): List<Issue>
    suspend fun findByParent(parentId: IssueId): List<Issue>
    suspend fun findByStatus(status: IssueStatus): List<Issue>
    suspend fun findByType(type: IssueType): List<Issue>
    suspend fun findByAssignee(assigneeId: String): List<Issue>
    suspend fun save(issue: Issue)
    suspend fun saveAll(issues: List<Issue>)
    suspend fun delete(id: IssueId)
    suspend fun exists(id: IssueId): Boolean
}